src/main/java/mate/academy/service/impl/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,104 @@
+package mate.academy.service.impl;
+
+import java.util.HashSet;
+import java.util.Optional;
+import lombok.RequiredArgsConstructor;
+import mate.academy.dto.shoppingcart.ShoppingCartDto;
+import mate.academy.exception.EntityNotFoundException;
+import mate.academy.mapper.ShoppingCartMapper;
+import mate.academy.model.Book;
+import mate.academy.model.CartItem;
+import mate.academy.model.ShoppingCart;
+import mate.academy.model.User;
+import mate.academy.repository.book.BookRepository;
+import mate.academy.repository.cartitem.CartItemRepository;
+import mate.academy.repository.shoppingcart.ShoppingCartRepository;
+import mate.academy.repository.user.UserRepository;
+import mate.academy.service.ShoppingCartService;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final BookRepository bookRepository;
+    private final CartItemRepository cartItemRepository;
+    private final UserRepository userRepository;
+
+    public ShoppingCartDto getShoppingCartForUser(Long userId) {
+        return shoppingCartMapper.toDto(shoppingCartRepository.findByUserId(userId));
+    }
+
+    public ShoppingCartDto addToCart(Long userId, Long bookId, int quantity) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId);
+        if (shoppingCart == null) {
+            shoppingCart = createNewShoppingCart(userId);
+        }
+
+        Book book = bookRepository.findById(bookId)
+                .orElseThrow(() -> new EntityNotFoundException("Book not found"));
+
+        Optional<CartItem> existingCartItem = shoppingCart.getCartItems().stream()
+                .filter(item -> item.getBook().getId().equals(bookId))
+                .findFirst();
+
+        ShoppingCart finalShoppingCart = shoppingCart;
```

Comment:
Why do you need this?

Reply:
decided to apply this
![image](https://github.com/user-attachments/assets/ff0dc013-2998-4502-853a-d57938764643)
i could split this part into two, first one will check the availability of the product in the cart, and the second to add a new product or update the quantity of an existing product, but then i need to use if-else instead of lambda. To my mind thats optimal solution